src/quartz/simulator/schedule.cpp

```diff
@@ -1006,11 +1004,52 @@ get_schedules(const CircuitSeq &sequence,
   std::unordered_map<std::string, std::queue<int>> gate_indices;
 
   for (auto &local_qubit : local_qubits) {
-    // convert vector<int> to vector<bool>
-    std::vector<bool> local_qubit_mask(num_qubits, false);
+    auto previous_local_qubit_mask = local_qubit_mask;
+    // Convert vector<int> to vector<bool>.
+    local_qubit_mask.assign(num_qubits, false);
     for (auto &i : local_qubit) {
       local_qubit_mask[i] = true;
     }
+    std::cout << std::endl;
+    if (current_local_qubit_layout.empty()) {
+      // First iteration. Take the initial layout from |local_qubits[0]|.
+      current_local_qubit_layout = local_qubit;
+    } else {
+      // Update the layout.
+      // We should have the same number of local qubits.
+      assert(local_qubit.size() == current_local_qubit_layout.size());
+      int num_global_swaps = 0;
+      for (auto &i : local_qubit) {
+        if (!previous_local_qubit_mask[i]) {
+          num_global_swaps++;
+        }
+      }
+      int should_swap_ptr = 0;
+      for (int i = (int)current_local_qubit_layout.size() - num_global_swaps;
+           i < (int)current_local_qubit_layout.size(); i++) {
+        if (!local_qubit_mask[current_local_qubit_layout[i]]) {
+          // This qubit should be swapped with a global qubit,
+          // and it is already at the correct position.
+          continue;
+        }
+        while (local_qubit_mask[current_local_qubit_layout[should_swap_ptr]]) {
+          should_swap_ptr++;
+        }
+        // Find the first local qubit that should be swapped with a global
+        // qubit, and perform a local swap.
+        std::swap(current_local_qubit_layout[i],
+                  current_local_qubit_layout[should_swap_ptr]);
+        should_swap_ptr++;
+      }
+      for (auto &i : local_qubit) {
+        if (!previous_local_qubit_mask[i]) {
+          // Put the new local qubit at its corresponding position.
+          current_local_qubit_layout[(int)current_local_qubit_layout.size() -
+                                     num_global_swaps] = i;
```

Comment:
So do we also need to keep track of the global qubits' mapping? Like,
```
0,1,2,3,4,5 | 6,7,8,9
4,1,2,6,7,9 | 3,0,8,5
```
Here the `3,0,5` comes from the 3 highest local qubits just before the local-global swapping, and `6,7,9` comes from the global qubits, both preserving the order?

Reply:
Yes, you're right. I think we need to track that, since it will affect the local layout for next iter. For example, if in next iter global qubits are [1, 2, 3, 0] then if current layout is [4,1,2,6,7,9 | 3,0,8,5] it will become [4,7,9,6,8,5|3,0,1,2], however if current layout is [4,1,2,9,6,7 | 3,0,8,5] (we do not preserve the order when doing global swaps in prev iter), it will become [4,6,7,9,8,5 | 3,0,1,2]. So we should also be consistent on how to do global swaps to make sure we always have the same local qubit layout for each stage.